File: virtual_mouse.py
import cv2
import mediapipe as mp
import pyautogui
import math
import win32gui
import win32con

# 获取屏幕尺寸
screen_width, screen_height = pyautogui.size()

# 导入手部检测任务
from mediapipe.tasks import python
from mediapipe.tasks.python import vision

# 初始化手部检测模型
base_options = python.BaseOptions(model_asset_path='hand_landmarker.task')
options = vision.HandLandmarkerOptions(
    base_options=base_options,
    num_hands=1,
    min_hand_detection_confidence=0.8,
    min_hand_presence_confidence=0.8,
    min_tracking_confidence=0.8
)

# 创建手部检测器
detector = vision.HandLandmarker.create_from_options(options)

# 打开摄像头
cap = cv2.VideoCapture(0)

# 设置摄像头帧率
cap.set(cv2.CAP_PROP_FPS, 30)

# 设置摄像头分辨率为640x480，提高处理速度
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)

# 点击状态
clicking = False

# 拖动状态
dragging = False

# 用于平滑鼠标移动的变量
mouse_history = []
history_length = 5  # 保存最近5次的鼠标位置

# 用于平滑手部关键点的变量
hand_landmarks_history = []
hand_history_length = 3  # 保存最近3次的手部关键点

# 创建窗口
cv2.namedWindow('Hand Tracking', cv2.WINDOW_NORMAL)

# 计算窗口大小为屏幕的三分之一，减少绘制开销
window_width = int(screen_width * 0.33)
window_height = int(screen_height * 0.33)

# 调整窗口大小和位置
cv2.resizeWindow('Hand Tracking', window_width, window_height)
# 将窗口居中显示
window_x = (screen_width - window_width) // 2
window_y = (screen_height - window_height) // 2
cv2.moveWindow('Hand Tracking', window_x, window_y)

# 定义中心操作区域（窗口的三分之二）
operation_margin_x = window_width // 6
operation_margin_y = window_height // 6
operation_x = operation_margin_x
operation_y = operation_margin_y
operation_width = window_width * 2 // 3
operation_height = window_height * 2 // 3

# 获取窗口句柄并设置为顶层窗口
hwnd = win32gui.FindWindow(None, 'Hand Tracking')
win32gui.SetWindowPos(hwnd, win32con.HWND_TOPMOST, 0, 0, 0, 0, win32con.SWP_NOMOVE | win32con.SWP_NOSIZE)

while True:
    # 读取摄像头画面
    ret, frame = cap.read()
    if not ret:
        break
    
    # 水平翻转画面，使操作更直观
    frame = cv2.flip(frame, 1)
    
    # 调整帧大小以匹配窗口大小
    frame = cv2.resize(frame, (window_width, window_height))
    
    # 创建一个透明背景的图像
    overlay = frame.copy()
    overlay = cv2.cvtColor(overlay, cv2.COLOR_BGR2BGRA)
    overlay[:, :, 3] = 0  # 设置alpha通道为0，完全透明
    
    # 获取画面尺寸
    frame_height, frame_width, _ = frame.shape
    
    # 将BGR转换为RGB
    rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    
    # 创建MediaPipe图像对象
    mp_image = mp.Image(image_format=mp.ImageFormat.SRGB, data=rgb_frame)
    
    # 检测手部
    results = detector.detect(mp_image)
    
    if results.hand_landmarks:
        for hand_landmarks in results.hand_landmarks:
            # 保存手部关键点到历史记录
            hand_landmarks_history.append(hand_landmarks)
            
            # 保持历史记录长度
            if len(hand_landmarks_history) > hand_history_length:
                hand_landmarks_history.pop(0)
            
            # 计算平滑后的手部关键点
            if len(hand_landmarks_history) > 1:
                # 创建一个新的列表来存储平滑后的关键点
                smoothed_landmarks = []
                for i in range(21):  # 21个关键点
                    # 计算所有历史记录中该关键点的平均值
                    avg_x = sum(lm[i].x for lm in hand_landmarks_history) / len(hand_landmarks_history)
                    avg_y = sum(lm[i].y for lm in hand_landmarks_history) / len(hand_landmarks_history)
                    avg_z = sum(lm[i].z for lm in hand_landmarks_history) / len(hand_landmarks_history)
                    
                    # 创建一个新的Landmark对象（这里使用一个简单的类来模拟）
                    class SmoothedLandmark:
                        def __init__(self, x, y, z):
                            self.x = x
                            self.y = y
                            self.z = z
                    
                    smoothed_landmarks.append(SmoothedLandmark(avg_x, avg_y, avg_z))
            else:
                # 如果历史记录不足，使用原始关键点
                smoothed_landmarks = hand_landmarks
            
            # 绘制手部连接
            connections = [
                (0, 1), (1, 2), (2, 3), (3, 4),  # 拇指
                (0, 5), (5, 6), (6, 7), (7, 8),  # 食指
                (0, 9), (9, 10), (10, 11), (11, 12),  # 中指
                (0, 13), (13, 14), (14, 15), (15, 16),  # 无名指
                (0, 17), (17, 18), (18, 19), (19, 20)   # 小指
            ]
            for connection in connections:
                start_idx, end_idx = connection
                start_point = smoothed_landmarks[start_idx]
                end_point = smoothed_landmarks[end_idx]
                # 计算相对于窗口的坐标
                start_x = int(start_point.x * window_width)
                start_y = int(start_point.y * window_height)
                end_x = int(end_point.x * window_width)
                end_y = int(end_point.y * window_height)
                # 绘制手部连接线，使用白色线条，更清晰
                cv2.line(overlay, (start_x, start_y), (end_x, end_y), (255, 255, 255, 255), 3)
            
            # 绘制手部关键点
            for landmark in smoothed_landmarks:
                # 计算相对于窗口的坐标
                x = int(landmark.x * window_width)
                y = int(landmark.y * window_height)
                # 绘制关键点，使用白色圆点
                cv2.circle(overlay, (x, y), 4, (255, 255, 255, 255), -1)
            
            # 获取平滑后的食指指尖坐标 (索引8)
            index_finger_tip = smoothed_landmarks[8]
            # 获取平滑后的拇指指尖坐标 (索引4)
            thumb_tip = smoothed_landmarks[4]
            # 获取平滑后的中指指尖坐标 (索引12)
            middle_finger_tip = smoothed_landmarks[12]
            
            # 计算食指在窗口中的坐标
            index_x = int(index_finger_tip.x * window_width)
            index_y = int(index_finger_tip.y * window_height)
            
            # 检查食指是否在中心操作区域内
            if operation_x <= index_x <= operation_x + operation_width and operation_y <= index_y <= operation_y + operation_height:
                # 将操作区域内的坐标映射到整个屏幕范围
                # 计算食指在操作区域内的相对位置 (0-1)
                rel_x = (index_x - operation_x) / operation_width
                rel_y = (index_y - operation_y) / operation_height
                
                # 应用轻微的非线性映射，提高中心区域的精度
                # 使用二次函数映射，使中心区域的移动更加精确
                rel_x = rel_x * rel_x * 2 if rel_x < 0.5 else 1 - (1 - rel_x) * (1 - rel_x) * 2
                rel_y = rel_y * rel_y * 2 if rel_y < 0.5 else 1 - (1 - rel_y) * (1 - rel_y) * 2
                
                # 将相对位置映射到整个屏幕
                screen_x = int(rel_x * screen_width)
                screen_y = int(rel_y * screen_height)
                
                # 添加到历史记录
                mouse_history.append((screen_x, screen_y))
                
                # 保持历史记录长度
                if len(mouse_history) > history_length:
                    mouse_history.pop(0)
                
                # 计算移动平均值，使用加权平均提高响应速度
                if mouse_history:
                    # 为最近的位置赋予更高的权重
                    weights = [i + 1 for i in range(len(mouse_history))]
                    total_weight = sum(weights)
                    
                    avg_x = sum(x * w for (x, y), w in zip(mouse_history, weights)) // total_weight
                    avg_y = sum(y * w for (x, y), w in zip(mouse_history, weights)) // total_weight
                    
                    # 检查是否处于拖动状态
                    if dragging:
                        # 拖动状态下，按住鼠标左键移动
                        pyautogui.dragTo(avg_x, avg_y, button='left', duration=0.05)
                    else:
                        # 正常状态下，移动鼠标
                        pyautogui.moveTo(avg_x, avg_y, duration=0.05)
                
                # 计算食指和拇指之间的距离（用于点击）
                index_thumb_distance = math.sqrt(
                    (index_finger_tip.x - thumb_tip.x) ** 2 +
                    (index_finger_tip.y - thumb_tip.y) ** 2 +
                    (index_finger_tip.z - thumb_tip.z) ** 2
                )
                
                # 计算拇指和中指之间的距离（用于拖动）
                thumb_middle_distance = math.sqrt(
                    (thumb_tip.x - middle_finger_tip.x) ** 2 +
                    (thumb_tip.y - middle_finger_tip.y) ** 2 +
                    (thumb_tip.z - middle_finger_tip.z) ** 2
                )
                
                # 当食指和拇指距离小于阈值时执行点击
                if index_thumb_distance < 0.05:
                    if not clicking:
                        pyautogui.click()
                        clicking = True
                else:
                    clicking = False
                
                # 当拇指和中指距离小于阈值时进入拖动状态
                if thumb_middle_distance < 0.05:
                    if not dragging:
                        dragging = True
                else:
                    dragging = False
            else:
                # 食指不在操作区域内，重置状态
                clicking = False
                dragging = False
    
    # 绘制中心操作区域的线框
    cv2.rectangle(overlay, (operation_x, operation_y), (operation_x + operation_width, operation_y + operation_height), (255, 255, 255, 200), 2)
    
    # 显示透明的手部追踪画面
    cv2.imshow('Hand Tracking', overlay)
    
    # 按退格键退出
    if cv2.waitKey(1) & 0xFF == 8:
        break

# 释放资源
cap.release()
cv2.destroyAllWindows()
detector.close()

File: virtual_mouse.cpp
#include <opencv2/opencv.hpp>
#include <mediapipe/framework/calculator_framework.h>
#include <mediapipe/framework/formats/image_frame.h>
#include <mediapipe/framework/formats/image_frame_opencv.h>
#include <mediapipe/solutions/hands/hands.h>
#include <mediapipe/solutions/hands/hands_options.pb.h>
#include <Windows.h>
#include <iostream>

using namespace cv;
using namespace mediapipe;
using namespace std;

// 全局变量
bool clicking = false;
bool dragging = false;

// 鼠标移动历史记录
vector<POINT> mouseHistory;
const int historyLength = 5;

// 手部关键点历史记录
vector<vector<NormalizedLandmark>> handLandmarksHistory;
const int handHistoryLength = 3;

// 屏幕尺寸
int screenWidth, screenHeight;

// 窗口尺寸
int windowWidth, windowHeight;

// 操作区域
Rect operationRegion;

// 计算两点之间的距离
float calculateDistance(const NormalizedLandmark& a, const NormalizedLandmark& b) {
    return sqrt(pow(a.x() - b.x(), 2) + pow(a.y() - b.y(), 2) + pow(a.z() - b.z(), 2));
}

// 平滑鼠标移动
POINT smoothMouseMove(const POINT& newPoint) {
    mouseHistory.push_back(newPoint);
    if (mouseHistory.size() > historyLength) {
        mouseHistory.erase(mouseHistory.begin());
    }

    if (mouseHistory.empty()) {
        return newPoint;
    }

    // 加权平均
    POINT avgPoint = { 0, 0 };
    int totalWeight = 0;
    for (int i = 0; i < mouseHistory.size(); i++) {
        int weight = i + 1;
        avgPoint.x += mouseHistory[i].x * weight;
        avgPoint.y += mouseHistory[i].y * weight;
        totalWeight += weight;
    }
    avgPoint.x /= totalWeight;
    avgPoint.y /= totalWeight;

    return avgPoint;
}

// 平滑手部关键点
vector<NormalizedLandmark> smoothHandLandmarks(const vector<NormalizedLandmark>& newLandmarks) {
    handLandmarksHistory.push_back(newLandmarks);
    if (handLandmarksHistory.size() > handHistoryLength) {
        handLandmarksHistory.erase(handLandmarksHistory.begin());
    }

    if (handLandmarksHistory.size() <= 1) {
        return newLandmarks;
    }

    vector<NormalizedLandmark> smoothedLandmarks;
    for (int i = 0; i < newLandmarks.size(); i++) {
        NormalizedLandmark landmark;
        float avgX = 0, avgY = 0, avgZ = 0;
        for (const auto& history : handLandmarksHistory) {
            avgX += history[i].x();
            avgY += history[i].y();
            avgZ += history[i].z();
        }
        avgX /= handLandmarksHistory.size();
        avgY /= handLandmarksHistory.size();
        avgZ /= handLandmarksHistory.size();
        landmark.set_x(avgX);
        landmark.set_y(avgY);
        landmark.set_z(avgZ);
        smoothedLandmarks.push_back(landmark);
    }

    return smoothedLandmarks;
}

// 映射坐标到屏幕
POINT mapToScreen(const Point& point) {
    // 计算相对位置
    float relX = static_cast<float>(point.x - operationRegion.x) / operationRegion.width;
    float relY = static_cast<float>(point.y - operationRegion.y) / operationRegion.height;

    // 非线性映射
    if (relX < 0.5) {
        relX = relX * relX * 2;
    } else {
        relX = 1 - pow(1 - relX, 2) * 2;
    }
    if (relY < 0.5) {
        relY = relY * relY * 2;
    } else {
        relY = 1 - pow(1 - relY, 2) * 2;
    }

    // 映射到屏幕
    POINT screenPoint;
    screenPoint.x = static_cast<int>(relX * screenWidth);
    screenPoint.y = static_cast<int>(relY * screenHeight);

    return screenPoint;
}

// 控制鼠标
void controlMouse(const POINT& point, bool isDragging) {
    if (isDragging) {
        // 拖动模式
        mouse_event(MOUSEEVENTF_LEFTDOWN, point.x, point.y, 0, 0);
        SetCursorPos(point.x, point.y);
        mouse_event(MOUSEEVENTF_LEFTUP, point.x, point.y, 0, 0);
    } else {
        // 正常模式
        SetCursorPos(point.x, point.y);
    }
}

// 执行点击
void performClick() {
    mouse_event(MOUSEEVENTF_LEFTDOWN, 0, 0, 0, 0);
    mouse_event(MOUSEEVENTF_LEFTUP, 0, 0, 0, 0);
}

int main() {
    // 获取屏幕尺寸
    RECT desktop;
    const HWND hDesktop = GetDesktopWindow();
    GetWindowRect(hDesktop, &desktop);
    screenWidth = desktop.right;
    screenHeight = desktop.bottom;

    // 计算窗口尺寸（屏幕的三分之一）
    windowWidth = screenWidth / 3;
    windowHeight = screenHeight / 3;

    // 计算操作区域（窗口的三分之二）
    int marginX = windowWidth / 6;
    int marginY = windowHeight / 6;
    operationRegion = Rect(marginX, marginY, windowWidth * 2 / 3, windowHeight * 2 / 3);

    // 初始化MediaPipe Hands
    HandsOptions handsOptions;
    handsOptions.set_max_num_hands(1);
    handsOptions.set_min_detection_confidence(0.8);
    handsOptions.set_min_tracking_confidence(0.8);
    unique_ptr<Hands> hands = Hands::Create(handsOptions);

    // 打开摄像头
    VideoCapture cap(0);
    if (!cap.isOpened()) {
        cerr << "无法打开摄像头" << endl;
        return -1;
    }

    // 设置摄像头参数
    cap.set(CAP_PROP_FRAME_WIDTH, 640);
    cap.set(CAP_PROP_FRAME_HEIGHT, 480);
    cap.set(CAP_PROP_FPS, 30);

    // 创建窗口
    namedWindow("Hand Tracking", WINDOW_NORMAL);
    resizeWindow("Hand Tracking", windowWidth, windowHeight);
    moveWindow("Hand Tracking", (screenWidth - windowWidth) / 2, (screenHeight - windowHeight) / 2);

    // 设置窗口为顶层
    HWND hWnd = FindWindowA(NULL, "Hand Tracking");
    SetWindowPos(hWnd, HWND_TOPMOST, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE);

    Mat frame;
    while (true) {
        // 读取摄像头画面
        cap >> frame;
        if (frame.empty()) {
            break;
        }

        // 水平翻转画面
        flip(frame, frame, 1);

        // 调整帧大小
        resize(frame, frame, Size(windowWidth, windowHeight));

        // 创建透明背景
        Mat overlay = Mat::zeros(frame.size(), CV_8UC4);

        // 转换为RGB
        Mat rgbFrame;
        cvtColor(frame, rgbFrame, COLOR_BGR2RGB);

        // 创建MediaPipe图像
        ImageFrame imageFrame(ImageFormat::SRGB, rgbFrame.cols, rgbFrame.rows, ImageFrame::kDefaultAlignmentBoundary);
        memcpy(imageFrame.MutablePixelData(), rgbFrame.data, rgbFrame.total() * rgbFrame.elemSize());

        // 检测手部
        auto results = hands->Process(imageFrame);

        if (results.multi_hand_landmarks().size() > 0) {
            const auto& handLandmarks = results.multi_hand_landmarks(0);
            
            // 平滑手部关键点
            auto smoothedLandmarks = smoothHandLandmarks(handLandmarks.landmark());

            // 绘制手部连接
            vector<pair<int, int>> connections = {
                {0, 1}, {1, 2}, {2, 3}, {3, 4},  // 拇指
                {0, 5}, {5, 6}, {6, 7}, {7, 8},  // 食指
                {0, 9}, {9, 10}, {10, 11}, {11, 12},  // 中指
                {0, 13}, {13, 14}, {14, 15}, {15, 16},  // 无名指
                {0, 17}, {17, 18}, {18, 19}, {19, 20}   // 小指
            };
            for (const auto& connection : connections) {
                int startIdx = connection.first;
                int endIdx = connection.second;
                const auto& startPoint = smoothedLandmarks[startIdx];
                const auto& endPoint = smoothedLandmarks[endIdx];
                Point start(int(startPoint.x() * windowWidth), int(startPoint.y() * windowHeight));
                Point end(int(endPoint.x() * windowWidth), int(endPoint.y() * windowHeight));
                line(overlay, start, end, Scalar(255, 255, 255, 255), 3);
            }

            // 绘制手部关键点
            for (const auto& landmark : smoothedLandmarks) {
                Point point(int(landmark.x() * windowWidth), int(landmark.y() * windowHeight));
                circle(overlay, point, 4, Scalar(255, 255, 255, 255), -1);
            }

            // 获取指尖坐标
            const auto& indexFingerTip = smoothedLandmarks[8];  // 食指
            const auto& thumbTip = smoothedLandmarks[4];       // 拇指
            const auto& middleFingerTip = smoothedLandmarks[12];  // 中指

            // 计算食指在窗口中的坐标
            Point indexPoint(int(indexFingerTip.x() * windowWidth), int(indexFingerTip.y() * windowHeight));

            // 检查食指是否在操作区域内
            if (operationRegion.contains(indexPoint)) {
                // 映射到屏幕坐标
                POINT screenPoint = mapToScreen(indexPoint);
                
                // 平滑鼠标移动
                POINT smoothedPoint = smoothMouseMove(screenPoint);
                
                // 控制鼠标
                controlMouse(smoothedPoint, dragging);

                // 计算距离
                float indexThumbDistance = calculateDistance(indexFingerTip, thumbTip);
                float thumbMiddleDistance = calculateDistance(thumbTip, middleFingerTip);

                // 点击检测
                if (indexThumbDistance < 0.05) {
                    if (!clicking) {
                        performClick();
                        clicking = true;
                    }
                } else {
                    clicking = false;
                }

                // 拖动检测
                if (thumbMiddleDistance < 0.05) {
                    dragging = true;
                } else {
                    dragging = false;
                }
            } else {
                // 重置状态
                clicking = false;
                dragging = false;
            }
        }

        // 绘制操作区域
        rectangle(overlay, operationRegion, Scalar(255, 255, 255, 200), 2);

        // 显示画面
        imshow("Hand Tracking", overlay);

        // 按退格键退出
        if (waitKey(1) == 8) {
            break;
        }
    }

    // 释放资源
    cap.release();
    destroyAllWindows();

    return 0;
}